### PR TITLE
1911 integrate new picture validator

### LINKED ISSUE #1911: Improve picture validation process
**LINKED WITH #1917**

The current picture validation process is not strict enough. The following sample does not produce any error.

```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. Pgm1.
       ENVIRONMENT DIVISION.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 a PIC X.
       01 b PIC whatever.
       PROCEDURE DIVISION.
           MOVE a TO b
           .
       END PROGRAM Pgm1.
```

The new class `TypeCobol.Compiler.Types.PictureValidator` created with the Semantic Domain does not validate the picture but does not produce any error message.

- Merge `TypeCobol.Compiler.Diagnostics.DataDescriptionChecker.CheckPicture` and new `PictureValidator` into a single validation code
- Ensure proper error messages are produced during the validation

### PR DESCRIPTION
**Pull request by fm-117**
Fixes #1911.

**Changes in this PR:**
- refactoring of `PictureValidator`
  - code reorganization
  - validation messages improved
- Integration into Code Element checker instead of Regex

This PR is incomplete, we still have to handle CURRENCY SIGN correctly, see issue #1987.

### CODE REVIEW

**Review by mayanje — DISMISSED** [review]

File: TypeCobol/Compiler/Nodes/Data.cs [hunk 1]
```diff
@@ -655,7 +644,8 @@ public string Hash
         }
 
         #region TypeProperties
-        public AlphanumericValue Picture { get {return _ComonDataDesc != null ? _ComonDataDesc.Picture : null;}}
+        public AlphanumericValue Picture => _ComonDataDesc?.Picture;
+        internal PictureValidator.Result PictureValidationResult => _ComonDataDesc?.PictureValidationResult;
         public bool IsJustified { get {  if(_ComonDataDesc != null && _ComonDataDesc.IsJustified != null) return _ComonDataDesc.IsJustified.Value; else return false; } }
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Nodes/Data.cs]: See c88ed087fe65dedd9176ad77926037dec948bbc2.

File: TypeCobol/Compiler/Types/PictureType.cs [hunk 2]
```diff
@@ -10,135 +10,58 @@ namespace TypeCobol.Compiler.Types
...
-
-        /// <summary>
-        /// String constructor
-        /// </summary>
-        /// <param name="value">Picture string value</param>
-        /// <param name="separateSign">a boolean value indicating whether the sign is separate character</param>
-        public PictureType(string value, bool separateSign) : this(new PictureValidator(value, separateSign))
-        {            
-        }
+        public static bool IsUsageCompatible(UsageFormat usage) =>
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Types/PictureType.cs]: Yes ! See c88ed087fe65dedd9176ad77926037dec948bbc2.

**Review by delevoye — APPROVED** [review]

**Review by mayanje — APPROVED** [review]

File: TypeCobol.Test/Parser/Programs/Cobol85/TypoParentheses.rdz.cbl [hunk 3]
```diff
@@ -57,12 +57,10 @@
...
        01 pic S9)5(V9(5.
-      * OK
        01 pic S9(5)V95.
        01 pic S95)V95.
        01 pic S9)5)V95.
        01 pic S9(5V95.
-      * OK
        01 pic S95V95.
        01 pic S9)5V95.
        01 pic S9(5(V95.
```
Comment (rooksdo) [inline TypeCobol.Test/Parser/Programs/Cobol85/TypoParentheses.rdz.cbl:66]: No errors detected on : 
```cob
01 pic zzzbb+zz-.
01 pic aa9b+.
01 pic $-***9b.
01 pic $v.
01 pic b9b$.
01 pic *9v-99.
01 pic ,,.
01 pic ,/.
01 pic 9/0*.
```
IBM compiler says : 
`"PICTURE"string character"Z"was out of sequence or was invalid with the preceding"PICTURE"string characters."PICTURE"string"S9(1)"was assumed.`
`"AA9B+"was not a valid combination of"PICTURE"string characters.  A"PICTURE"string of"S9(1)"was assumed.`
`"PICTURE"string character"*"was out of sequence or was invalid with the preceding"PICTURE"string characters."PICTURE"string"S9(1)"was assumed.`
`"$V"was not a valid combination of"PICTURE"string characters.  A"PICTURE"string of"S9(1)"was assumed.`
`"PICTURE"string character"$"was out of sequence or was invalid with the preceding"PICTURE"string characters."PICTURE"string"S9(1)"was assumed.`
`"PICTURE"string character"9"was out of sequence or was invalid with the preceding"PICTURE"string characters."PICTURE"string"S9(1)"was assumed.`
`",,"was not a valid combination of"PICTURE"string characters.  A"PICTURE"string of"S9(1)"was assumed.`
`",/"was not a valid combination of"PICTURE"string characters.  A"PICTURE"string of"S9(1)"was assumed.`
`"PICTURE"string character"*"was out of sequence or was invalid with the preceding"PICTURE"string characters."PICTURE"string"S9(1)"was assumed.`

**Review by rooksdo — CHANGES_REQUESTED** [review]

File: TypeCobol.Test/Parser/Programs/Cobol85/TypoParentheses.rdz.cbl (hunk 3 above)
Comment (fm-117) [reply TypeCobol.Test/Parser/Programs/Cobol85/TypoParentheses.rdz.cbl:66]: See new tests.

**Comment by fm-117** [comment]
The automata has been completely rewritten taking code from [1911_IntegrateNewPictureValidator_JCM](https://github.com/TypeCobolTeam/TypeCobol/tree/1911_IntegrateNewPictureValidator_JCM) as reference.

**Review by mayanje — APPROVED** [review]

**Review by delevoye — APPROVED** [review]

**Comment by fm-117** [comment]
Closed to be recreated (because of pending change request from @rooksdo).

**Pull request closed without merging**